### DISCUSSION TITLE
Clarifying max_tokens Usage and Limits in OpenAI API (Issue #3195 Update)

### DISCUSSION THREAD

**Discussion by Algorithm5838** [ChatGPTNextWeb/NextChat#3208] [General]
After reviewing OpenAI's documentation, I now have a clearer understanding of the `max_tokens` parameter. Initially, when I opened an issue #3195, I had misunderstood its function. I've also noticed that there may be a similar misunderstanding reflected in the description of `max_tokens` in the settings:
> Max Tokens
> Maximum length of input tokens and generated tokens 
___
### Terminology:
- `input` = `input tokens` = `prompt` = `message` — These terms are synonymous and refer to **the user's prompt**.
- `output` = `generated tokens` = `completion` — These terms are synonymous and refer to **the AI's response**.
- `context window` = `context length` — These terms are synonymous and refer to **the model's context**.

### Clarifications:
- According to OpenAI's [API Reference](https://platform.openai.com/docs/api-reference/chat/create#completions/create-max_tokens), `max_tokens` pertain only to the `output`:
> The maximum number of tokens to generate in the chat **completion**.
- As stated by OpenAI in their descriptions of the models `gpt-4-1106-preview`, `gpt-4-vision-preview`, and `gpt-3.5-turbo-1106` ([source](https://platform.openai.com/docs/models/gpt-4-and-gpt-4-turbo)), `max_tokens`, as of date, should not exceed `4096`:
> Returns a maximum of 4,096 output tokens.
- The sum of `input` and `max_tokens` must not surpass the `context window`, or an error will occur (refer to the first link for details):
> The total length of input tokens and generated tokens is limited by the model's context length.
- Users can reduce the `max_tokens` to allocate more tokens for `input`, with the trade-off being a shorter AI response, as it will be confined to the set number of `max_tokens`.
___
With this understanding, we can discern the causes of the errors in my previous issue.

### Error 1, with model `gpt-3.5-turbo-1106`:
```json
{
  "error": {
    "message": "max_tokens is too large: 8192. This model supports at most 4096 completion tokens, whereas you provided 8192.",
    "type": "invalid_request_error",
    "param": "max_tokens",
    "code": null
  }
}
```
The error arose because our requested `max_tokens` of `8192` exceeded the model's maximum allowed `max_tokens` of `4096`, as noted above.

### Error 2, with model `gpt-4`:
```json
{
  "error": {
    "message": "This model's maximum context length is 8192 tokens. However, you requested 8256 tokens (64 in the messages, 8192 in the completion). Please reduce the length of the messages or completion.",
    "type": "invalid_request_error",
    "param": "messages",
    "code": "context_length_exceeded"
  }
}
```
This error occurred because the sum of the `max_tokens` of `8192` and the input tokens requested exceeded the model's maximum context length of `8192`, as explained previously.
___
### Conclusion:
In implementation, the maximum value for `max_tokens` should not be higher than `context window`-1, for example `4095`.
There are two ways to go about it, in my opinion, for the general user, but I don't know how to code it:
1. General (easy): Calculating `max_tokens` for the users by dividing `context window` by 2. For example: if `context window` was `4096`, `max_token` would be `2048`, and so on. It should not exceed this, the user though should only be able to decrease it.
2. Dynamic: Calculating `man_tokens` for the users by using this formula: (`max_tokens` = `context window` - `input tokens`). We already know the value of `context window`, since it is chosen by the user, we just need to calculate `input tokens` and do the subtraction.
I prefer the first option.

Note: An additional good approach for the first option, would be to introduce a checkbox with a warning, allowing informed users to opt-in to increasing the `max_tokens` value if they understand the implications. The warning could be something like: 
> Warning: The sum of input tokens and max_tokens must not exceed the model's context window.
Increasing max_tokens would force you to decrease input tokens, and decreasing it would allow you to use more input tokens.

Either way, I find it useful if the user could know the value of tokens of their prompts during writing it and after, and the AI's tokens.

All this should be considered if you want to control the length of the AI's response, or if there were models that require it.

**Comment by H0llyW00dzZ** [comment]
it will break if there is lots messages in chats because we had this 

![image](https://github.com/Yidadaa/ChatGPT-Next-Web/assets/17626300/9e5a6d60-0a06-44e5-982a-37e822680e44)

and this for prevent hallucination 

![image](https://github.com/Yidadaa/ChatGPT-Next-Web/assets/17626300/4546a57a-d0c4-41f3-9ba3-2cef03be68d6)

**Reply by Algorithm5838** [reply]
I will rewrite the conclusion. Check it again after, and tell me your thoughts.

**Reply by H0llyW00dzZ** [reply]
alright 
let me know because its confusing using max_tokens for old models

**Reply by Algorithm5838** [reply]
Done.

You use max_tokens to control the length of the AI's tokens in its response. Memory prompt and system prompt are not included in max_tokens. An error would occur not because of max_tokens, but because the sum of input tokens and max_tokens exceeded the context window.

This is the formula to remember: input_tokens = context window - max_tokens

For example, if you are using `gpt-3.5-turbo` now, you should not use `4096` in max_tokens, because you didn't leave yourself tokens for `input`. input tokens (x) = context window (4096) - max_tokens (4096) -> x = 0
But if you used `2048` for max_tokens, you would have only `2048` tokens for `input`.

**Reply by H0llyW00dzZ** [reply]
memory prompt and system prompt use a max_tokens from model which is selected
if we set `2048` it's not enough for user input

**Reply by Algorithm5838** [reply]
As I have understood it, I don't believe so, but I don't know how to check it.
And I disagree because that is what I understood from OpenAI's definition of max_tokens:
> The maximum number of tokens to generate in the chat completion.

Which is also what I found while searching community.openai.com.
And per the error message, completion is the same value you use in max_tokens, which, per the terminology I laid out before, is the length of the AI's response.

Regarding `2048`, the user would have `2048` tokens for input. If the user wants the output to be the same length as his input, they should leave it as is, and if it was not enough, they should opt for the new model `gpt-3.5-turbo-1106`, which has the same price and more tokens (16k), or they can opt for the old 16k models or other models. But if they only want more tokens in input, they should decrease max_tokens, if they increased it they would have less input tokens. That is the problem with the 4k models.

You can see the problem of 4k models with this example:
I asked ChatGPT to translate two chapters (3166 tokens) in French into English (meaning input is more than 2048), it stopped before completing chapter 1, mid-sentence (859 tokens). But if max_tokens was `2048`, at least the user would know that this model is not capable of handling this text, since input + output would be around 6200 tokens.

**Reply by H0llyW00dzZ** [reply]
Well, The Inject system prompts and memory prompts use a model selected by the user. In a scenario where we set 'max_tokens' (as user) in the payload to, for instance, '2048', it would work. However, memory prompts, also known as summarizing, randomly consume a significant amount of tokens. This could potentially lead to a multitude of errors in the future.

**Reply by Algorithm5838** [reply]
I just tested the system message, and as per the following error, it is not part of max_tokens, like I said earlier:
Note: I set max_tokens as `4096`
```json
{
  "error": {
    "message": "This model's maximum context length is 4097 tokens. However, you requested 7335 tokens (3239 in the messages, 4096 in the completion). Please reduce the length of the messages or completion.",
    "type": "invalid_request_error",
    "param": "messages",
    "code": "context_length_exceeded"
  }
}
```
I used the two chapters (3166 tokens) as system message, and just sent "hello". You can see that it is included part of `messages` = `input` and not part of `completion` = `output` = `max_tokens`. If the system message was part of `completion` = `output` = `max_tokens`, there should be no error.

So, Inject system prompts, won't affect max_tokens = AI's response, and I guess Memory prompt would be the same. They would be part of input tokens, with the user's prompt.

**Reply by H0llyW00dzZ** [reply]
nah it will affect max_tokens which can break conversation
and memory prompt is summarizing your conversation (example)

**Reply by Algorithm5838** [reply]
Your point is valid in the sense that any tokens used in the conversation, including system and memory prompts, count towards the total token count that must not exceed the model's context window limit. However, these are part of the input tokens, not the max_tokens which are reserved for the AI's response.
At least, this is my understanding from the sources that I read.

**Reply by H0llyW00dzZ** [reply]
Yes, it's also not currently possible to fix this issue, such as through refactoring or adding `max_tokens`, because it would break the old models cause `context_length_exceeded` errors.

**Reply by Algorithm5838** [reply]
Yeah, maybe it is better to leave it as is.
There will always be max_tokens, even if we haven't set it, because OpenAI will automatically calculate it themselves: `max_tokens = context window - input`. You can test it by sending 4000 tokens in a 4k model, and the AI will generate less than 100 tokens in response.

**Reply by H0llyW00dzZ** [reply]
Yes, the only thing that requires `max_tokens` is for GPT-4-Vision
